Backends/CLX/clipboard.lisp

```diff
@@ -0,0 +1,249 @@
+                  ;;     value of the SetSelectionOwner request to the timestamp of
+                  ;;     the event triggering the acquisition attempt, not to
+                  ;;     CurrentTime. A zero-length append to a property is a way to
+                  ;;     obtain a timestamp for this purpose; the timestamp is in
+                  ;;     the corresponding PropertyNotify event.
+                  ;;
+                  ;; The reasons for his seems to be to ensure that the ownership is
+                  ;; actually transferred correctly. This shouldn't be a major issue
+                  ;; in practice, and it significantly simplifies the implementation.
+                  (xlib:set-selection-owner (xlib:window-display window) selection window nil)
+                  (if (eq (xlib:selection-owner (xlib:window-display window) selection) window)
+                      (call-next-method)
+                      (warn "Couldn't set X11 selection ~s owner to ~s." selection window))))))
+  (frob :primary)
+  (frob :clipboard))
+
+(macrolet ((frob (selection)
+             `(defmethod climb:release-selection ((port clx-selection-mixin)
+                                                  (selection (eql ,selection))
+                                                  (object climb:selection-object))
+                (when-let ((window (sheet-direct-xmirror (climb:selection-object-owner object))))
+                  (xlib:set-selection-owner (xlib:window-display window) selection nil nil))
+                (call-next-method))))
+  (frob :primary)
+  (frob :clipboard))
+
+;;;  Paste support
+;;;
+;;;  When a client requests the content of the clipboard or selection, a REQUESTED-OBJECT
+;;;  parameter is used to indicate what kind of of object the caller expects to
+;;;  receive. Selection objects are explained in ports.lisp.
+;;;
+;;;  First we shortcut result if the selection owner is a sheet in the same image. Then we
+;;;  try to negotiate the type – ask for :TARGETS. If response is empty we default to
+;;;  :STRING. Then we verify whether we can translate proposed types to the requested one
+;;;  and if it is possible we request that type with convert-selection. Otherwise NIL is
+;;;  returned.
+
+(defun clx-request-selection (port selection requested-object)
+  ;; Shortcut the request if the selection is owned by a window in the same
+  ;; image. This improves translations and overall responsiveness.
+  (setf (clipboard-outstanding-request port) requested-object)
+  (let ((sheet (selection-object-owner requested-object))
+        (window (sheet-xmirror (selection-object-owner requested-object)))
+        (to-type (selection-object-type requested-object))
+        (table (climi::get-object-table requested-object)))
+    (labels ((wait-for-request ()
+               (or (not (clipboard-outstanding-request port))
+                   (not (eq selection (selection-object-content requested-object)))))
+             (negotiate-selection-type ()
```

Comment:
The negotiation has a problem. If a string is requested, this will result in a target of `:STRING`. This is incorrect. After requesting the available types (using a `:TARGETS` request), the "best" type should be requested:

If `:UTF8_STRING` is available, it should be used.
If only `:STRING` is available, then it should be used, but decoded using ISO-8859-1.

Doing a bit of tracing shows that the current implementation uses `:TEXT`, should should never be offered at all.

The list of string types that should be offered is different from the list of types that McCLIM should accept.

Reply:
fixed